### ISSUE TITLE
Java wrapper script

### ISSUE THREAD

**Issue by bgruening** [bioconda/bioconda-recipes#1519]
Currently we use a bash script to wrap Java programs ins Bioconda. This is a port to python of the same script. I think this is more maintainable and we can define a strict dependency on python. Whereas the old wrapper depends on bash, which is not often available in minimal Docker containers and there is no dependency yet (alternatively we could write one).

I'm not sure this is worth the effort and defining python as dependency for a Java program also looks wired. On the other side this script should be more portable and maintainable.

ping @chapmanb @bioconda/core 

```
#!/usr/bin/env python
#
# Wrapper script for Java Conda packages that ensures that the java runtime
# is invoked with the right options. Adapted from the bash script (http://stackoverflow.com/questions/59895/can-a-bash-script-tell-what-directory-its-stored-in/246128#246128).
#

#
# Program Parameters
#

jar_file = 'PeptideShaker-1.1.3.jar'

default_jvm_mem_opts = ['-Xms512m', '-Xmx1g']

# !!! End of parameter section. No user-serviceable code below this line !!!


import sys
from os import access, getenv, system, X_OK
from os.path import dirname, realpath, join as join_path
from subprocess import call


def real_dirname(path):
    """Return the symlink-resolved, canonicalized directory-portion of path."""
    return dirname(realpath(path))


def java_executable():
    """Return the executable name of the Java interpreter."""
    java_home = getenv('JAVA_HOME')
    java_bin = join_path('bin', 'java')

    if java_home and access(join_path(java_home, java_bin), X_OK):
        return join_path(java_home, java_bin)
    else:
        return 'java'


def jvm_opts(argv):
    """Construct list of Java arguments based on our argument list.

    The argument list passed in argv must not include the script name.
    The return value is a 3-tuple lists of strings of the form:
      (memory_options, prop_options, passthrough_options)
    """
    mem_opts = []
    prop_opts = []
    pass_args = []

    for arg in argv:
        if arg[0:2] == '-D':
            prop_opts.append(arg)
        elif arg[0:3] == '-XX':
            prop_opts.append(arg)
        elif arg[0:3] == '-Xm':
            mem_opts.append(arg)
        else:
            pass_args.append(arg)

    # In the original shell script the test coded below read:
    #   if [ "$jvm_mem_opts" == "" ] && [ -z ${_JAVA_OPTIONS+x} ]
    # To reproduce the behaviour of the above shell code fragment
    # it is important to explictly check for equality with None
    # in the second condition, so a null envar value counts as True!

    if mem_opts == [] and getenv('_JAVA_OPTIONS') == None:
        mem_opts = default_jvm_mem_opts

    return (mem_opts, prop_opts, pass_args)


def main():
    java = java_executable()
    jar_dir = real_dirname(sys.argv[0])
    (mem_opts, prop_opts, pass_args) = jvm_opts(sys.argv[1:])

    if pass_args != [] and pass_args[0][0:2] == 'eu':
        jar_arg = '-cp'
    else:
        jar_arg = '-jar'

    jar_path = join_path(jar_dir, jar_file)

    java_args = [java]+ mem_opts + prop_opts + [jar_arg] + [jar_path] + pass_args
    java_args = ' '.join(java_args)

    sys.exit(call(java_args, shell=True))


if __name__ == '__main__':
    main()
```

**Comment by chapmanb** [comment]
Björn;
Cool, thanks for doing this. python code looks so much better than bash. The idea behind the bash scripts was that they're portable outside of conda so folks can take and use them on most systems, but this is just as portable in a conda environment.

The main issue I see is that the errors from the wrapped program are going to be confusing since you'll get a python stacktrace surrounding the actual error. In my experience this will point people in the wrong direction and make it harder to identify the underlying problems. I'm not sure if any of the more lightweight python system call approaches avoid this.

**Comment by bgruening** [comment]
I guess they do, but they are not as portable as `subprocess`. os.system() has problems with the return values - not sure this matters so much. Let me know what you think and I will proceed or abandon this idea :)

**Comment by johanneskoester** [comment]
Interesting. Thanks for working on this!! I think this is a good idea since Python comes for free with conda. We should just make sure that this works with all Python versions. A few comments:

* You can wrap `call()` with a try except and hide any python stacktrace in case of an error.
* There is no need to join the command line args, call can take a list directly.
* Is there a particular reason to use `shell=True`?
* The slicing of arg in `jvm_opts` can be replaced with `str.startswith` instead.
* I think it is good practice to not import functions from os and os.path by name. Instead, just use the full path to the function, e.g. `os.path.exists`. For me, this looks more readable.

**Comment by dakl** [comment]
The main issue with capturing any errors and presenting them to the user is that some tools have a non-zero exit status even if no "real error" has occurred. An example of this is when `picard` is invoked without any parameters. The user is shown a list of subcommands, and the return code is 1. 

~~~
$ picard
USAGE: PicardCommandLine <program name> [-h]

Available Programs:
--------------------------------------------------------------------------------------
Fasta:                                           Tools for manipulating FASTA, or related data.
    CreateSequenceDictionary                     Creates a SAM or BAM file from reference sequence in fasta format
...
    VcfFormatConverter                           Converts a VCF file to a BCF file, or BCF to VCF
    VcfToIntervalList                            Converts a VCF or BCF file to a Picard Interval List.

--------------------------------------------------------------------------------------

$ echo $? 
1
~~~

This means that if this is wrapped with a check on the exit code, the user will be presented with an error (from python) even though the tool developer didn't mean for an error to be thrown (there is no java stack trace). 

This is obviously a problem with the individual tools, in this case picard. If no error is intended, the return code should be 0. However I think it's too common that developer mis-handle exit codes to make a python wrapper feasible atm, at least without displaying some python stack traces to the user...  :(

**Comment by dakl** [comment]
Fyi; https://github.com/broadinstitute/picard/pull/553

**Comment by dakl** [comment]
Ok then, I was clearly confused when I wrote my previous comment. I've cleaned up the script and put it here: 

https://github.com/dakl/bioconda-recipes/blob/java-python-wrapper/scripts/java-wrapper.py

Let me know what you think.

**Comment by johanneskoester** [comment]
Looks good! I wonder what would be a good place for now. Maybe put it into scripts/ and add an information about how to use it in the Java section of the guidelines.

**Comment by chapmanb** [comment]
Daniel;
Thanks for this. I tested it out and was wrong about the python stacktraces. I don't see them for failed commands with just a subprocess.call wrapped in the sys.exit. Great -- sorry for throwing up an incorrect red flag. The only thing I noticed were two typos of 'startwith` instead of `startswith` on lines 55 and 77. Thanks again to both of you for working on this. Agree that scripts is a good place and we can test it out on a few java formula to see how it does.
